scipy/optimize/_tstutils.py

```diff
@@ -1,61 +1,678 @@
+     randomly negative on the other.
+     f4 - f6 are not continuous at the root.
+
+  'aps': The test problems in the 1995 paper
+     TOMS "Algorithm 748: Enclosing Zeros of Continuous Functions"
+     by Alefeld, Potra and Shi.  Real-valued functions of
+     real-valued inputs on an interval with a zero.
+     Suitable for methods which start with an enclosing interval, and
+     derivatives up to 2nd order.
+
+  'complex': Some complex-valued functions of complex-valued inputs.
+     No enclosing bracket is provided.
+     Suitable for methods which use one or more starting values, and
+     derivatives up to 2nd order.
+
+  The test cases are provided as a list of dictionaries. The dictionary
+  keys will be a subset of:
+  ["f", "fprime", "fprime2", "args", "bracket", "smoothness",
+  "a", "b", "x0", "x1", "root", "ID"]
+"""
+
+# Sources:
+#  [1] Alefeld, G. E. and Potra, F. A. and Shi, Yixun,
+#      "Algorithm 748: Enclosing Zeros of Continuous Functions",
+#      ACM Trans. Math. Softw. Volume 221(1995)
+#       doi = {10.1145/210089.210111},
+
 from random import random
 
+import numpy as np
+
 from scipy.optimize import zeros as cc
 
+# "description" refers to the original functions
+description = """
+f2 is a symmetric parabola, x**2 - 1
+f3 is a quartic polynomial with large hump in interval
+f4 is step function with a discontinuity at 1
+f5 is a hyperbola with vertical asymptote at 1
+f6 has random values positive to left of 1, negative to right
+
+of course these are not real problems. They just test how the
+'good' solvers behave in bad circumstances where bisection is
+really the best. A good solver should not be much worse than
+bisection in such circumstance, while being faster for smooth
+monotone sorts of functions.
+"""
+
 
 def f1(x):
-    return x*(x-1.)
+    r"""f1 is a quadratic with roots at 0 and 1"""
+    return x * (x - 1.)
+
+
+def f1_fp(x):
+    return 2 * x - 1
+
+
+def f1_fpp(x):
+    return 2
 
 
 def f2(x):
-    return x**2 - 1
+    r"""f2 is a symmetric parabola, x**2 - 1"""
+    return x ** 2 - 1
+
+
+def f2_fp(x):
+    return 2 * x
+
+
+def f2_fpp(x):
+    return 2
 
 
 def f3(x):
-    return x*(x-1.)*(x-2.)*(x-3.)
+    r"""A quartic with roots at 0, 1, 2 and 3"""
+    return x * (x - 1.) * (x - 2.) * (x - 3.)  # x**4 - 6x**3 + 11x**2 - 6x
+
+
+def f3_fp(x):
+    return 4 * x ** 3 - 18 * x ** 2 + 22 * x - 6
+
+
+def f3_fpp(x):
+    return 12 * x ** 2 - 36 * x + 22
 
 
 def f4(x):
+    r"""Piecewise linear, left and right discontinuous at x=1, the root."""
     if x > 1:
-        return 1.0 + .1*x
+        return 1.0 + .1 * x
     if x < 1:
-        return -1.0 + .1*x
+        return -1.0 + .1 * x
     return 0
 
 
 def f5(x):
+    r"""Hyperbola with a pole at x=1, but pole replaced with 0.  Not continuous at root."""
     if x != 1:
-        return 1.0/(1. - x)
+        return 1.0 / (1. - x)
     return 0
 
 
+# f6(x) returns random value. Without memoization, calling twice with the
```

Comment:
Is this useful? The general rule is never to use a random number without seeding the generator; any test failure won't be reproducible.

Reply:
Actually this "function" is a real pain for many reasons :-( 
(Perhaps that is why it is there...)
`f6` appears to have been present since at least 2002 (commit ba5207cabf41db45bf03f70b0724f3e7e376a60a)
I added the memoization for reproducibility within a single root-finding session, never-mind agreement across platforms or time.

Reply:
Ah well, it doesn't fail occasionally, so in practice not a problem. I commented on it just because it doesn't look quite right, but let's leave as is.

> Actually this "function" is a real pain for many reasons :-(
> (Perhaps that is why it is there...)

:)